### ISSUE TITLE
document.write!! and goog.require() problem 

### ISSUE THREAD

**Issue by azizzaeny** [google/closure-library#844]
in targeted browser environment `goog.require` after document.load is not allowed because google closure use writeScriptTag which is violent and it wipe out the page and throw error 'Cannot write "' + src + '" after document load'. 
```js
// line 1232 of base.js
  goog.writeScriptTag_ = function(src, opt_sourceText) {
    if (goog.inHtmlDocument_()) {
      /** @type {!HTMLDocument} */
      var doc = goog.global.document;

      // If the user tries to require a new symbol after document load,
      // something has gone terribly wrong. Doing a document.write would
      // wipe out the page. This does not apply to the CSP-compliant method
      // of writing script tags.
      if (!goog.ENABLE_CHROME_APP_SAFE_SCRIPT_LOADING &&
          doc.readyState == 'complete') {
        // Certain test frameworks load base.js multiple times, which tries
        // to write deps.js each time. If that happens, just fail silently.
        // These frameworks wipe the page between each load of base.js, so this
        // is OK.
        var isDeps = /\bdeps.js$/.test(src);
        if (isDeps) {
          return false;
        } else {
          throw Error('Cannot write "' + src + '" after document load');
        }
      }
...
```

we asume that dependency graph is not changes after page load, and it had been done compiled at the first build step, then evaluated all the way down

but in dynamic environment and interactive development ... dependencies are changes a lot.. and we communicate with `goog.require` to pull out new dependencies..

this is not posible because `goog.require` use `document.write('<script url>')`, we need to find better way to bring js into evaluated environment without document.write 

for example:
```js
function loadScript(sScriptSrc,callbackfunction) {
    var oHead = document.getElementsByTagName('head')[0];
    var oScript = document.createElement('script');
    oScript.setAttribute('src',sScriptSrc); 
    oScript.setAttribute('type','text/javascript');
    var loadFunction = function() {
      if (this.readyState == 'complete' || this.readyState == 'loaded') {
        callbackfunction();
      }
    };
    oScript.onreadystatechange = loadFunction;
    oScript.onload = callbackfunction;
    oHead.appendChild(oScript);
  }
```
```js
loadScript(host + ":" + port + "/socket.io/socket.io.js", init);
```
what was evaluated may need changes, using chrome console repl and call `goog.require` to pull dependencies and being able to evaluate things after page load, for me it is huge impact on productivity, 

our `goog.require` should be able to load after page loaded, what do you think? :+1:

**Comment by shicks** [comment]
@jplaisted is working on revamping our debug loader to work with ES6 modules, and part of that involves making it more asynchronous.  I believe this will come for free when that rolls out, though I can't say for sure.

**Comment by jplaisted** [comment]
Things would be more asynchronous but not necessarily dynamic. I don't think we really want to support dynamic loading post document load. Even with a new debug loader for ES6 modules you'd still have to hack things. We can't resolve relative paths post script evaluation. So you'd have to use absolute paths or symbols/namespaces. Later is fine but the former breaks the conventions we're adding (granted azizzaeny is doing in his example, so maybe users are fine with it, but it'd be confusing, imo). Also the debug loader doesn't get loaded in compiled or bundled environments, meaning you can't dynamic load there.

Back over to @shicks to decide if this something we want to support in the future. But the new debug loader I'm writing for ES6 modules will not add this; we'd have to address the above concerns and then probably some more.

**Comment by shicks** [comment]
I guess I misunderstood the request.  We are looking into the possibility of building a late-loader into Closure, but it's unclear what its mechanism will be, and it's a lot less certain to happen.  As @jplaisted points out, it's unlikely we'll want to tie it to the same symbols used by `goog.require` (especially since these are increasingly dinosaurs in our ES6 world).  So I'd say stay tuned and please chime in on a separate issue if you have thoughts about late loading, but this particular issue is WAI.

**Comment by shicks** [comment]
Just to be clear, we consider late loading an orthogonal issue to `goog.require` and to the ES6 module loading issue, and it's honestly somewhat low priority.

In particular is that we're primarily concerned with late-loading production (compiled) code, and it's pretty unclear how this would fit into Closure Library.  It's not impossible that we'd work on it, but I wouldn't hold me breath on it.

**Comment by azizzaeny** [comment]
the hack was using `goog.net.loadMany`

```js
goog.provide = goog.exportPath_
goog.lobal.CLOSURE_IMPORT_SCRIPT = function(file){
 if(goog.inHtmlDocument_){  // loaded
    goog.net.jsloader.load(file)
  }
}
```
use somewhere goog.net.jsloader.loadMany()`, will pull dependencies

but now its deprecated, @shicks and @jplaisted do you know any alternative to `goog.net.jsloader` ?

the examples above are just plain, rough and doesn't enough solution but the idea is there..

we can actually call `goog.require` after page loaded it will produce null
with requirement the first `goog.addDependencies` graph is introduced and the first goog.require call is evaluated
but when the first `goog.require` not yet evaluated then the problem is there..

to reproduce my issue add goog/base and goog/deps.js to html
```html
<script src="public/main.out/goog/base.js"></script>
<script src="public/main.out/goog/deps.js"></script>
```
then you eval in REPL `goog.require('somens.foo')` to invoke `document.write` at `document.body` it wil refuse because first `goog.require` must introduced before `document.ready`

when you add at require before document.ready
```html
<script src="public/main.out/goog/base.js"></script>
<script src="public/main.out/goog/deps.js"></script>
<!-- others js -->
<script>goog.require("somens.foo");</script>
```
then you eval then you eval in REPL `goog.require('somens.bar') it will ok and output null and the dependencies scripts is not there meaning not loaded

namespaces to path ns->path `foo.bar` mean `foo/bar.js` right?
based relative to base.js  `http://<<server>>/goog/base.js`
`goog.require('foo.bar')` => eval `load_script("foo/bar.js")` =>  `http://<<server>>/foo/bar.js`

the idea of `goog.require` is to __ensure dependencies is there__


i come along this issue when using clojurescript
i cannot use `goog.module` because cljs generated google-closure

in cljs when you declare namespaces `(ns foo.bar)` it would produce
`goog.provide('foo.bar')`   -> produce global code
`goog.require('cljs.core')` -> ensure core loaded

in es6
`goog.module` --> ensure loaded as module es6 -> translated into `goog.loadModule((exports)=> )`

i think the issue was seperate each other, es6 has its own mechanism of async code loading and plain es5 must have its own instead document.write

and hopefully there is no dinosours in `goog.require` ;-)

**Comment by azizzaeny** [comment]
Although all of this are not necessarily
 when compilation bundle is advanced.
this is to support other user that develop code incrementally

lets define whats is __interactive and dynamic development__ (not dynamic loading, i think we had already dynamic loading)

1. __interactive__ mean incrementall evaluation
2. __dynamic__ mean reintroduce new code in runtime without long compiling code again

what i do was just connect to browser REPL with editor and then start evaluating code there _(changing things and live patching there)_

interactive mean we evaluate code from repl and sends it through the evaluation envionment `[browser/nodejs]`

when you have ES6 REPL it mean you `[type ES6 code] --> [transform to es5] --> [browser understand es5 then evaluate it ]`
when you have cljs repl `[write cljs code] --> [transform to es5] -> [browser understand that] -> [evaluate inside the browser]`

long compilation can be skipped by just translating what necesarly to load in browser or in nodejs

yes future is must asynchronously and also future development should be supported

future development is when you connected to your creations and changes things inside but to do that at the compensate you must build it onces, once it compiled load it in evironment 

connected:
`[compile] -> [evaluated dynamicly]`
common approach:
`[compile es6] -> [changes] -> [reload] -> [compile again] -> [reload]`

`goog.require` need its own asynchronous loading not necesarly like es6 debug loader (nb: i think this is theissue too in es6 world when the code must be compiled to understand it)

we just dont do document.write!! maybe replace it by head.write!! or xhr.eval in dev-mode google-compiler

if we call in cljs `(require '[cljsjs.react])` => it will translated and function call `(loadScirpt "cljsjs/react.js")`
when we call it again and again incrementally  `(require '[cljsjs.react])` `(require '[cljsjs.react])` => `loadScript` is skipped because its already there
but when you have document.write it will have side effects that it will keep adding to the document.body

we can actually make it pure by remove previous same script.src and add again with `document.parrent.removeChild` 

the issue is small and tiny fix, replace document.write with somethings else in dev mode , in production advanced compilation they all bundled 

can we reopen the case maybe others will have same issue and have difference perspective?

**Comment by shicks** [comment]
It sounds like you're saying you're not actually interested in late loading, *per se*, but that for some reason you still need to do your initial code loading *after* document.ready.

I agree that we want to support the case of loading uncompiled code directly in the browser without a long compilation step.  Where you're losing me is that I don't understand why this doesn't work for you.  I think you're saying it has to do with wanting to use `goog.require` for "live patching", i.e. replacing (a subset of) JS implementation in place without reloading the entire HTML.

I understand that there are Chrome developer tools to do this, but have never looked into how they work.  I'm not sure that `goog.require` is appropriate for this - for example, and already-existing instances of old classes will no longer be `instanceof` the new replaced version of the same class - you need some deep browser integration to make that work right.

Or am I still misunderstanding what you're trying to do?

I don't know the complete history, but my understanding is that we use document.write for a reason - specifically, we need to work in CSP contexts where eval is not allowed.  So simply replacing one for the other is a non-starter.  Maybe there's a way for you to switch into the eval mode using the already existing infrastructure?

**Comment by joeychia** [comment]
I also meet the similar issue. My application FOO is post loading another script (BAR.js) which uses goog.require() to acquire all its dependencies. The BAR.js is unknown before FOO's onload event. Since goog.require() is using document.write after pageload, it wipes out the entire DOM.
I'm looking for a way to avoid this. Maybe some parameter for goog.require() to append dependencies as <script> element to existing DOM (like body).

**Comment by shicks** [comment]
TL;DR: I think you might be able to get what you need by bootstrapping the following before loading `base.js`:

```js
var CLOSURE_DEFINES = {'goog.ENABLE_CHROME_APP_SAFE_SCRIPT_LOADING': true};
```

Would you mind trying that out?

Longer story:

I find the following in `base.js`:

```js
/**
 * @define {boolean} Whether to use a Chrome app CSP-compliant method for
 *     loading scripts via goog.require. @see appendScriptSrcNode_.
 */
goog.define('goog.ENABLE_CHROME_APP_SAFE_SCRIPT_LOADING', false);

  /**
   * Appends a new script node to the DOM using a CSP-compliant mechanism. This
   * method exists as a fallback for document.write (which is not allowed in a
   * strict CSP context, e.g., Chrome apps).
   *
   * NOTE: This method is not analogous to using document.write to insert a
   * <script> tag; specifically, the user agent will execute a script added by
   * document.write immediately after the current script block finishes
   * executing, whereas the DOM-appended script node will not be executed until
   * the entire document is parsed and executed. That is to say, this script is
   * added to the end of the script execution queue.
   *
   * The page must not attempt to call goog.required entities until after the
   * document has loaded, e.g., in or after the window.onload callback.
   *
   * @param {string} src The script URL.
   * @private
   */
goog.appendScriptSrcNode_ = function(src) {
    /** @type {Document} */
    var doc = goog.global.document;
    var scriptEl =
        /** @type {HTMLScriptElement} */ (doc.createElement('script'));
    scriptEl.type = 'text/javascript';
    scriptEl.src = src;
    scriptEl.defer = false;
    scriptEl.async = false;
    doc.head.appendChild(scriptEl);
  };

  /**
   * The default implementation of the import function. Writes a script tag to
   * import the script.
   *
   * @param {string} src The script url.
   * @param {string=} opt_sourceText The optionally source text to evaluate
   * @return {boolean} True if the script was imported, false otherwise.
   * @private
   */
  goog.writeScriptTag_ = function(src, opt_sourceText) {
    if (goog.inHtmlDocument_()) {
      /** @type {!HTMLDocument} */
      var doc = goog.global.document;

      // If the user tries to require a new symbol after document load,
      // something has gone terribly wrong. Doing a document.write would
      // wipe out the page. This does not apply to the CSP-compliant method
      // of writing script tags.
      if (!goog.ENABLE_CHROME_APP_SAFE_SCRIPT_LOADING &&
          doc.readyState == 'complete') {
        // Certain test frameworks load base.js multiple times, which tries
        // to write deps.js each time. If that happens, just fail silently.
        // These frameworks wipe the page between each load of base.js, so this
        // is OK.
        var isDeps = /\bdeps.js$/.test(src);
        if (isDeps) {
          return false;
        } else {
          throw new Error('Cannot write "' + src + '" after document load');
        }
      }

      if (opt_sourceText === undefined) {
        if (!goog.IS_OLD_IE_) {
          if (goog.ENABLE_CHROME_APP_SAFE_SCRIPT_LOADING) {
            goog.appendScriptSrcNode_(src); // uses document.appendChild
          } else {
            goog.writeScriptSrcNode_(src); // uses document.write
          }
        } else {
          // ... old IE code here ...
        }
      } else {
        doc.write(
            '<script type="text/javascript">' +
            goog.protectScriptTag_(opt_sourceText) + '</' +
            'script>');
      }
      return true;
    } else {
      return false;
    }
  };
```

So from this it looks to me that enabling this define should switch you to the codepath you need.  Potentially it's still missing some way to hook into a callback once it's loaded (though that can be worked around in your loaded script if necessary)?

**Comment by joeychia** [comment]
@shicks Yes. It works! Thank you!

**Comment by shicks** [comment]
Marking this as closed.
